debezium-sink/src/main/java/io/debezium/bindings/kafka/KafkaDebeziumSinkRecord.java

```diff
@@ -95,91 +244,177 @@ public boolean isTombstone() {
+                final Schema keySchema = keySchema();
+                if (keySchema == null) {
+                    throw new ConnectException("Configured primary key mode 'record_key' cannot have null schema");
                 }
-                case RECORD_VALUE -> {
-                    final Schema valueSchema = originalKafkaRecord.valueSchema();
-                    if (valueSchema != null && Schema.Type.STRUCT.equals(valueSchema.type())) {
-                        Struct afterPayload = getPayload();
-                        if (primaryKeyFields.isEmpty()) {
-                            return afterPayload;
-                        }
-                        else {
-                            return buildRecordValueKeyStruct(afterPayload, primaryKeyFields);
-                        }
-                    }
-                    else {
-                        throw new ConnectException("No struct-based primary key defined for record value.");
-                    }
+                else if (keySchema.type().isPrimitive()) {
+                    return keyAsStruct();
+                }
+                if (Schema.Type.STRUCT.equals(keySchema.type())) {
+                    return filterFields(keyAsStruct(), topicName(), allowedPrimaryKeyFields, fieldsFilter);
                 }
-                case RECORD_HEADER -> {
-                    final SchemaBuilder headerSchemaBuilder = SchemaBuilder.struct();
-                    originalKafkaRecord.headers().forEach((Header header) -> headerSchemaBuilder.field(header.key(), header.schema()));
-                    final Schema headerSchema = headerSchemaBuilder.build();
-                    final Struct headerStruct = new Struct(headerSchema);
-                    originalKafkaRecord.headers().forEach((Header header) -> headerStruct.put(header.key(), header.value()));
-                    return headerStruct;
+                else {
+                    throw new ConnectException("An unsupported record key schema type detected: " + keySchema.type());
                 }
             }
+            case RECORD_VALUE -> {
+                final Struct payload = getPayload();
+                final Schema valueSchema = payload.schema();
+                if (valueSchema != null && Schema.Type.STRUCT.equals(valueSchema.type())) {
+                    return filterFields(payload, topicName(), allowedPrimaryKeyFields, fieldsFilter);
+                }
+                else {
+                    throw new ConnectException("No struct-based primary key defined for record value");
+                }
+            }
+            case RECORD_HEADER -> {
+                if (originalKafkaRecord.headers().isEmpty()) {
+                    throw new ConnectException("Configured primary key mode 'record_header' cannot have empty message headers");
+                }
+                return getRecordHeaders();
+            }
+            case KAFKA -> {
+                return kafkaCoordinates();
+            }
```

Comment:
What do you think about adding a default handler in the switch that raises an exception so that if we add any new primary key modes in the future, we don't mistakenly cause NullPointerExceptions?

Reply:
Done